src/tools/imgtool/library.cpp

```diff
@@ -2,89 +2,68 @@
 // copyright-holders:Nathan Woods
 /****************************************************************************
 
-    library.c
+    library.cpp
 
     Code relevant to the Imgtool library; analgous to the MESS/MAME driver
     list.
 
 ****************************************************************************/
 
 #include <string.h>
+#include <algorithm>
 
 #include "imgtool.h"
 #include "library.h"
 #include "pool.h"
 
-struct imgtool_library
-{
-	object_pool *pool;
-	imgtool_module *first;
-	imgtool_module *last;
-};
+namespace imgtool {
 
 
+//-------------------------------------------------
+//  ctor
+//-------------------------------------------------
 
-imgtool_library *imgtool_library_create(void)
+library::library()
 {
-	imgtool_library *library;
-	object_pool *pool;
-
-	/* create a memory pool */
-	pool = pool_alloc_lib(nullptr);
-	if (!pool)
-		goto error;
-
-	/* allocate the main structure */
-	library = (imgtool_library *)pool_malloc_lib(pool, sizeof(imgtool_library));
-	if (!library)
-		goto error;
-
-	/* initialize the structure */
-	memset(library, 0, sizeof(*library));
-	library->pool = pool;
-
-	return library;
-
-error:
-	if (pool)
-		pool_free_lib(pool);
-	return nullptr;
+	m_pool = pool_alloc_lib(nullptr);
+	if (!m_pool)
+		throw std::bad_alloc();
 }
 
 
+//-------------------------------------------------
+//  dtor
+//-------------------------------------------------
 
-void imgtool_library_close(imgtool_library *library)
+library::~library()
 {
-	pool_free_lib(library->pool);
+	pool_free_lib(m_pool);
 }
 
 
+//-------------------------------------------------
+//  add_class
+//-------------------------------------------------
 
-static void imgtool_library_add_class(imgtool_library *library, const imgtool_class *imgclass)
+void library::add_class(const imgtool_class *imgclass)
 {
-	imgtool_module *module;
 	char *s1, *s2;
 
-	/* allocate the module and place it in the chain */
-	module = (imgtool_module *)imgtool_library_malloc(library, sizeof(*module));
+	// allocate the module and place it in the chain
+	m_modules.emplace_back(std::unique_ptr<imgtool_module>(new imgtool_module));
```

Comment:
Will `std::make_unique` not work here?  you need to be careful about using constructor syntax like that because in some cases it isn't exception-safe.